### PR TITLE
Extended jinja template to generate template-type semantic attributes.

### PR DESCRIPTION
**Pull request by AlexanderWert**
With https://github.com/open-telemetry/build-tools/pull/186, template-type attributes (such as `http.request.header.<key>`) can be defined in semantic convention model yaml files. With that, these template attributes can be used in code generation for the semantic conventions.

This PR adds a new class `AttributeKeyTemplate` for this kind of attribute keys and extends the code generation to cover attributes like `http.request.header.<key>`, `db.elasticsearch.path_parts.<key>`, `rpc.grpc.request.metadata.<key>`, etc.

With this change utility classes like  [`CapturedHttpHeadersUtil`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/CapturedHttpHeadersUtil.java#L36) in the java-instrumentation repo can be replaced with instances (and corresponding constants) of the more generic `AttributeKeyTemplate` class.

### CODE REVIEW

**Comment by AlexanderWert** [comment]
@open-telemetry/java-approvers PTAL, I think this would allow to replace some custom helpers in the agent with generated constants for this type of attributes

**Comment by trask** [comment]
cc @breedx-splk @mateuszrzeszutek to review new public API surface if you have a chance

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:59]: I have mixed feelings about encoding the `.replace('-', '_')` part here; it was recently removed from the HTTP headers semconv. Perhaps doing any sort of normalization should be the responsibility of the caller of this class?

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:17]: WDYT about a single static cache for all the templates?

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:15]: Can you add some javadocs to the public methods of this class?

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (breedx-splk) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:17]: Yeah, and then prefix the map key with the actual prefix?

**Review by breedx-splk — APPROVED** [review]
This looks good to me. Just have the one idea around the key cache being static, but otherwise seems like a nice addition.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:17]: Yeah, exactly that.

Comment (jack-berg) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:17]: What's the advantage of single map? Sounds like more contention and more resizing outside of calls for a particular prefix. I.e. if i call `HTTP_REQUEST_HEADER.getAttributeKey(..)` many times with distinct names but `RPC_GRPC_RESPONSE_METADATA.getAttributeKey(..)` for a small set of keys, the resizing of the shared map triggered by the cardinality of `HTTP_REQUEST_HEADER` could impact `RPC_GRPC_RESPONSE_METADATA`.

I suppose using a single map minimizes the memory footprint of initializing many maps, most of which won't get used for a particular application. But we could solve that by initializing the maps with a small initial capacity.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (jack-berg) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:59]: I agree. We don't normalize attribute keys elsewhere, and this shouldn't be the exception.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (AlexanderWert) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:17]: I'm fine with any of the above proposed approaches.

> I suppose using a single map minimizes the memory footprint of initializing many maps, most of which won't get used for a particular application. But we could solve that by initializing the maps with a small initial capacity.

Another alternative would be to do lazy initialization of the individual maps, to reduce the impact of template attributes that are not being used.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (trask) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:19]: I believe this would need to be `volatile` for thread safety. I'd suggest staying with `final` and inline initialization.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (trask) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java]: > With a single map/cache we would need to cast the AttributeKey objects, though, when retrieving.

this seems like a good enough reason to stick with instance field

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (jack-berg) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:19]: Technically I think this is one of those weird situations where it doesn't actually matter if two threads read this as null and instantiate / use different instances of `ConcurrenHashMap`, since the consequence is just that you will have created two copies of the same `AttributeKey` instead of returning a cached copy.

The inline initialization is simpler though and doesn't raise flags about thread safety. I'd say stick with the `final` inline initialization and keep this type of thing in our back pocket if we get feedback about utilization.

**Comment by AlexanderWert** [comment]
@jack-berg , @trask , @mateuszrzeszutek , @breedx-splk 

Based on the discussions above I reverted the implementation to use a final , inline-initialized map (with initial size 1) again. 
PTA(nother)L, Thanks!

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java]: > > With a single map/cache we would need to cast the AttributeKey objects, though, when retrieving.
> 
> this seems like a decent reason to stick with instance field

That's a good point. I don't have strong feelings about that, I think the current version is okay too.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:15]: Can you add JavaDoc on the class level too?

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (mateuszrzeszutek) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:59]: I have somewhat mixed feelings about lowercasing the key here. We did that for the HTTP headers, but there were some discussions about whether we should keep or ignore the case; and maybe, for other template attributes in the future we'll want to keep the case. I find that unlikely though -- I'm guessing we'll probably follow the precedent set by HTTP headers; and `key` and `Key` being different things is _super_ confusing.

**Review by mateuszrzeszutek — APPROVED** [review]
LGTM, I left one more minor JavaDoc comment.
Thanks!

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (AlexanderWert) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:59]: Yeah, I was basically following the HTTP headers example. But agree that for others we might need to keep the key as is. What do others think?

Comment (AlexanderWert) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:59]: Actually let me change that to not lowercasing it here and do the `toLowerCase()` in the instrumentation when using it for HTTP headers. WDYT?

Comment (mateuszrzeszutek) [reply src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java]: > Actually let me change that to not lowercasing it here and do the `toLowerCase()` in the instrumentation when using it for HTTP headers. WDYT?

👍 
Should be a bit more future-proof in case the semconv decides to use case sensitive naming somewhere (I hope they don't 😅 )

**Comment by AlexanderWert** [comment]
OK, then I think it's ready to be merged.

File: src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java
Comment (jack-berg) [inline src/main/java/io/opentelemetry/semconv/AttributeKeyTemplate.java:35]:
```suggestion
  public static AttributeKeyTemplate<String> stringKeyTemplate(String prefix) {
```
What do folks think about making these public? As is, they can only be used for attributes in `ResourceAttributes` and `SemanticAttributes`, but its reasonable for a user to want to define their own custom attribute template. Maybe that suggests that ultimately this class should be promoted to `io.opentelemetry.api.common` to live alongside `AttributeKey` and `Attributes`.

**Review by jack-berg — APPROVED** [review]
Couple nits and a question about whether the methods to create `AttributeKeyTemplate` should be public to allow users to create their own templates. None of the comments should block this PR.

**Pull request merged**
